src/main/scala/org/deepdive/ddlog/DeepDiveLogSemanticChecker.scala

```diff
@@ -142,6 +149,94 @@ object DeepDiveLogSemanticChecker extends DeepDiveLogHandler {
     }
   }
 
+  // collect variables used in the expression
+  def collectVariables(expr: Expr) : Set[String] = {
+    var set = new HashSet[String]()
```

Comment:
Can you try to eliminate these `var set` from these collectVariables functions?  I think it's unnecessarily complicating things.  Instead of `+= x`, `foreach`, and sequence of recursive calls, you can use `Set(x)`, `flatMap`, `++` to directly return the set of variables found for the `e`.